### PR TITLE
Integrated STP with Category Support

### PR DESCRIPTION
**Pull request by Spikey84**
This PR integrates @4z0t's ui mod called Specific Target Priorities to the game. This mod allows a player to select with a hotkey the kind of unit they with their units to target. In addition to integrating the current mod i am also making modifications that will allow the targeting of all variations of units (different factions) without selecting each on individually.

### CODE REVIEW

**Comment by Spikey84** [comment]
Things to do here:
* Add default TP hotkey
* Make hotkey look cleaner
* Finish/rework the category table
* check with @Garanas that performance is not effected by this change

**Comment by MrRowey** [comment]
image
Set Hotkey to Units to P and this is my error

**Comment by Spikey84** [comment]
@MrRowey do u have ATP and/or STP?

File: lua/keymap/misckeyactions.lua [hunk 1]
```diff
@@ -390,6 +390,73 @@ function SetWeaponPriorities(prioritiesString, name, exclusive)
...
+function findPriority(bpID)
+    local bp = __blueprints[bpID]
+    local categories = bp.Categories
+
+    local tech
+    local faction
+    local type
+    local field
+
+    for _, c in categories do
```
Comment (4z0t) [inline lua/keymap/misckeyactions.lua:428]: Use hashed categories to check

File: lua/keymap/misckeyactions.lua [hunk 2]
```diff
@@ -390,6 +390,73 @@ function SetWeaponPriorities(prioritiesString, name, exclusive)
...
+    faction = "categories." .. faction
+    type = "categories." .. type
+    field = "categories." .. field
+
+    local tp = tech .. " * " .. faction .. " * " .. type .. " * " .. field
+    local tp2 = tech .. " * " .. type .. " * " .. field
+    local tp3 = type .. " * " .. field
+    local tp4 = field
+
+    local masterTP = "{categories." .. bpID .. ", " .. tp .. ", " .. tp2 .. ", " .. tp3 .. ", " .. tp4 .. "}"
```
Comment (4z0t) [inline lua/keymap/misckeyactions.lua:456]: String format

File: lua/keymap/keydescriptions.lua [hunk 3]
```diff
@@ -455,4 +455,6 @@ keyDescriptions = {
     ['t3_support_naval_factory'] = '<LOC key_desc_0387>build T3 Support Naval Factory',
 
     ['recheck_targets_of_weapons'] = 'Recheck targets of weapons of selected units',
+    ['set_target_priority'] = 'Set Target Priority',
```
Comment (Garanas) [inline lua/keymap/keydescriptions.lua:458]: Make normal english sentences - No Additional Capitals

File: lua/keymap/misckeyactions.lua [hunk 4]
```diff
@@ -390,6 +390,73 @@ function SetWeaponPriorities(prioritiesString, name, exclusive)
     SimCallback({Func = 'WeaponPriorities', Args = {SelectedUnits = unitIds, prioritiesTable = priotable, name = name, exclusive = exclusive or false }})
 end
 
+function SetWeaponPrioritiesSpecific()
```
Comment (Garanas) [inline lua/keymap/misckeyactions.lua:393]: Please add a description to the function, as the title is not that specific (even though it is suggesting that it is 😄 )

File: lua/keymap/misckeyactions.lua [hunk 5]
```diff
@@ -390,6 +390,73 @@ function SetWeaponPriorities(prioritiesString, name, exclusive)
...
+        if text then
+            text = "\n" .. text
+        else
+            text = "\n" .. LOC(__blueprints[bpId].Interface.HelpText)
+        end
+        SetWeaponPriorities(findPriority(bpId), text, false)
+    end
+end
+
+function SetDefaultWeaponPriorities()
```
Comment (Garanas) [inline lua/keymap/misckeyactions.lua:408]: Please add a description to the function

File: lua/keymap/misckeyactions.lua (hunk 2 above)
Comment (Garanas) [reply lua/keymap/misckeyactions.lua:456]: For all of them 😄

**Review by Garanas — CHANGES_REQUESTED** [review]
The suggested changes still need some fine tuning, but look good in general. Functionally it appears to work already and I am already a huge fan 👍

**Comment by Garanas** [comment]
You can set the target priority to units of yourself, which isn't necessarily a bad thing but just wanted to point it out.

image
The code needs some fail safes (for modded units, in particular) when a certain field does not exist. Take the command unit, it has no tech level.

**Comment by Garanas** [comment]
As a few examples:

```
{ -- Shield disruptor
  categories.dal0310, 
  categories.TECH3 * categories.AEON * categories.DIRECTFIRE * categories.LAND, 
  categories.TECH3 * categories.DIRECTFIRE * categories.LAND, 
  categories.DIRECTFIRE * categories.LAND, 
  categories.LAND
}

{ -- Janus
  categories.dea0202, 
  categories.TECH2 * categories.UEF * categories.ANTIAIR * categories.AIR, 
  categories.TECH2 * categories.ANTIAIR * categories.AIR, 
  categories.ANTIAIR * categories.AIR, 
  categories.AIR
}

{ -- Bouncer
  categories.dal0310, 
  categories.TECH3 * categories.AEON * categories.DIRECTFIRE * categories.LAND, 
  categories.TECH3 * categories.DIRECTFIRE * categories.LAND, 
  categories.DIRECTFIRE * categories.LAND, 
  categories.LAND
}
```

The logic can be found in the changes, but roughly:
- unit id that we hover over
- common categories that the unit we hover over may share with similar units, including faction
- common categories that the unit we hover over may share with similar units
- common categories that the unit we hover over may share with similar units of any tech
- common categories that the unit we hover over may share with similar units of the same layer

**Review by Garanas — APPROVED** [review]

**Pull request merged**